walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // If there are no attendees in the request, return the entire day
+    if (request.getAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+      
+    // If duration of the request is longer than a day, return an empty array
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // List of all unavailable times
+    ArrayList<TimeRange> unavailableTimes = new ArrayList<>();
+
+    // Get list of attendees on meeting request
+    Collection<String> reqAttendees = request.getAttendees();
+    for (String reqAttendee : reqAttendees) {
+
+      // Check whether attendee is scheduled for other events
+      for (Event event : events) {
+        if (event.getAttendees().contains(reqAttendee)) {
+
+          // Add time range of event to list of unavailable time ranges
+          unavailableTimes.add(event.getWhen());
+        }
+      }
+    }
+
+    // If there are no conflicts, return the entire day
+    if (unavailableTimes.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Sort list of unavailable times from earliest to latest
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    // Remove nested time ranges and combine overlapping time ranges in list
+    int i = 0;
+    while (i < unavailableTimes.size()-1) {
+      TimeRange currTimeRange = unavailableTimes.get(i);
+      TimeRange nextTimeRange = unavailableTimes.get(i+1);
+
+      if (currTimeRange.contains(nextTimeRange)) {
+        unavailableTimes.remove(nextTimeRange);
+      } else if (nextTimeRange.contains(currTimeRange)) {
+        unavailableTimes.remove(currTimeRange);
+      } else if (currTimeRange.overlaps(nextTimeRange)) {
+        TimeRange newTimeRange = 
+            TimeRange.fromStartEnd(currTimeRange.start(), nextTimeRange.end(), false);
+        unavailableTimes.remove(currTimeRange);
+        unavailableTimes.remove(nextTimeRange);
+        unavailableTimes.add(i, newTimeRange);
+      } else {
+        i++;
```

Comment:
This is an interesting approach to the problem of skipping an element after removing one!

You could also consider using a `ListIterator` which I just discovered today: https://docs.oracle.com/javase/8/docs/api/java/util/ListIterator.html